---
title: "Bios CPH685 Homework 1"
subtitle: Due Sep 18, 2019 @ 11:59PM
output: 
      html_document:
        toc: true
        toc_depth: 4
#bibliography: /Users/jzhou/GoogleDrive/LatexBib/prop.bib
#csl: ../apa.csl
---
# Q1. Git/GitHub

**No handwritten homework reports are accepted for this course.** We work with Git and GitHub. Efficient and abundant use of Git, e.g., frequent and well-documented commits, is an important criterion for grading your homework.

1. Apply for the [Student Developer Pack](https://education.github.com/pack) at GitHub using your UA email.

2. Create a **private** repository `bios-cph685-2019-fall` (or something makes sense to you) and add `jinjinzhou` as your collaborators with write permission.

3. Top directories of the repository should be `hw1`, `hw2`, ... Maintain two branches `master` and `develop`. The `develop` branch will be your main playground, the place where you develop solution (code) to homework problems and write up report. The `master` branch will be your presentation area. Submit your homework files (R markdown file `Rmd`, `html` file converted from R markdown, all code and data sets to reproduce results) in `master` branch.

4. After each homework due date, I will check out your master branch for grading. Tag each of your homework submissions with tag names `hw1`, `hw2`, ... Tagging time will be used as your submission time. That means if you tag your `hw1` submission after deadline, penalty points will be deducted for late submission.

# Q2. Linux Shell Commands

The `35.230.38.116:/share/hw1` folder contains a typical genetic data set in plink format. If interested, you can read plink documentation at <http://zzz.bwh.harvard.edu/plink/>. But it's definitely not necessary for this homework.

- `merge-geno.bim` contains information of each genetic marker (SNP). Each line is a SNP and has 6 fields:  
`Chromosome`, `SNP ID`, `Genetic Distance (morgan)`, `Base Pair Position (bp)`, `Allele 1`, `Allele 2`.
    ```{bash,eval=T}
    head /share/hw1/merge-geno.bim
    ```

- `merge-geno.fam` contains individual information. Each line is one individual and has 6 fields:  
`Family ID`, `Person ID`, `Father ID`, `Mother ID`, `Sex` coded as 1 (male) or 2 (female), `Affection Status`  
`Father ID = 0` means that person's father is not in this data set. Similarly `Mother ID` = 0 means that person's mother is not in this data set.
    ```{bash,eval=T}
    head -20 /share/hw1/merge-geno.fam
    ```

- `merge-geno.bed` contains genotypes of each individual in binary format. We don't need this file for this homework.

Please, do **not** put these data files into Git; they are huge. You even don't need to copy them into your directory. Just read from the data folder `/share/hw1` directly.

Use Linux shell commands to answer following questions. 

1. How many persons are in the data set (statisticians call this `n`)? How many SNPs are in the data set (statisticians call this `p`)?  

**Solution:**  

```{bash, eval = TRUE}
wc -l /share/hw1/merge-geno.fam
wc -l /share/hw1/merge-geno.bim
```

There are 959 persons and 8348674 SNPS in the data set.  


2. Which chromosomes does this data set contain? How many SNPs are in each chromosome?  

**Solution:**

The number of SNPs in each chromosome will be listed as follows  

```{bash}
awk -F '\t' '{print $1}' /share/hw1/merge-geno.bim | sort | uniq -c | sort -nr

awk -F '\t' '{print $1}' /share/hw1/merge-geno.bim | sort | uniq -c | wc -l
```

There are 11 chromosomes in the data set  

3. MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp -- 48,130,769 bp. How many SNPs are located within MAP4 gene?  

**Solution:**  

```{bash}
awk -F '\t' '{ if(($1 == 3) && ($4 >= 47892180) && ($4 <= 48130769)) { print } }' /share/hw1/merge-geno.bim | wc -l
```


4. Statistical geneticists often have to reformat a data set to feed into various analysis programs. For example, to use the Mendel software <http://www.genetics.ucla.edu/software/mendel>, we have to reformat the data set to be read by Mendel.

      - Mendel's SNP definition file is similar to the plink `bim` file but has format  
      `SNP ID`, `Chromosome`, `Base Pair Position`  
      with each field separated by a comma. Write a Linux shell command to convert `merge-geno.bim` to Mendel SNP definition file. The first few lines of the Mendel SNP definition file should look like
    ```{bash, echo=FALSE, eval=F}
    head mendel_snpdef.txt
    ```
    
    - Mendel's pedigree file is similar to the plink `fam` file but has format  
    `Family ID`, `Person ID`, `Father ID`, `Mother ID`, `Sex` coded as M or F, `Twin Status`  
    with each field separated by a comma. Write a Linux shell command to convert `merge-geno.fam` to Mendel pedigree file. Since twin status is not available in plink format, we put nothing for that field. Also Mendel limits Person ID to have length less than or equal to 8 characters, so we have to strip the string `T2DG` from the IDs. First few lines of the Mendel pedigree should look like
    
    ```{bash, echo=FALSE, eval=F}
    head -20 mendel_ped.txt
    ```

**Solution:**  

```{bash}
awk -F '\t' '{ print $2 "," $1 "," $4}' /share/hw1/merge-geno.bim | head 
```

```{bash}
awk '{ 
gsub("T2DG", "", $2);
gsub("T2DG", "", $3);
gsub("T2DG", "", $4);
if ($5 == 1) $5 = "M";
else $5 = "F"; 
print $1 "," $2 "," $3 "," $4 "," $5 "," ""}' /share/hw1/merge-geno.fam | head
```

# Q3. R Batch Run

In class we discussed using R to organize simulation studies. 

1. Expand the `runSim.R`  script in D2L to include arguments `seed` (random seed), `n` (sample size), `dist` (distribution) and `rep` (number of simulation replicates). When `dist="gaussian"`, generate data from standard normal; when `dist="t1"`, generate data from t-distribution with degree of freedom 1 (same as Cauchy distribution); when `dist="t5"`, generate data from t-distribution with degree of freedom 5. Calling `runSim.R` will (1) set random seed according to argument `seed`, (2) generate data according to argument `dist`, (3) compute the primed-indexed average estimator in class and the classical sample average estimator for each simulation replicate, (4) report the average mean squared error (MSE). Notice that all three distributions have true mean `0`. 
$$
  \frac{\sum_{r=1}^{\text{rep}} (\widehat \mu_r - \mu_{\text{true}})^2}{\text{rep}}
$$
for both methods.  

**Solution:**  

```{r Installing R packages, echo = TRUE, eval = FALSE}
install.packages("pacman", repos = "http://cran.us.r-project.org")
```

```{r Loading R packages, echo = TRUE, eval = TRUE, results = "hide"}
require(pacman)
p_load(tidyverse, plyr, kableExtra)
```

```{r Loading Source File Into R}
source("runSim.R")
Sample.MSE(seed = 280, 
           n=100, 
           dist = "gaussian", 
           rep = 50)
Sample.MSE(seed = 280, 
           n=100, 
           dist = "t1", 
           rep = 50)
Sample.MSE(seed = 280, 
           n=100, 
           dist = "t5", 
           rep = 50)
```



2. Modify the `autoSim.R` script in D2L to run simulations with combinations of sample sizes `nVals = seq(100, 500, by=100)` and distributions `distTypes = c("gaussian", "t1", "t5")` and write output to appropriately named files. Use `rep = 50`, and `seed = 280`.  

**Solution:**

In this question, we do not modify the file since The `Sample.MSE.R` file can give you the final result as table shown in the following question.  


3. Write an R script to collect simulation results from output files and print average MSEs in a table of format

| $n$ | Method   | $t_1$ | $t_5$ | Gaussian |
|-----|----------|-------|-------|----------|
| 100 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 200 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 300 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 400 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 500 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |

**Solution:**

```{r, echo = TRUE, eval = TRUE, results = "asis"}
distTypes <- c("gaussian", "t1", "t5")
nVals <- seq(100, 500, by=100)
rep <- 50
seed <- 280

Params <- expand.grid(seed, nVals, distTypes, rep)
names(Params) <- c("seed", "n", "dist", "rep")

Res <- mdply(Params, Sample.MSE)
Res <- Res %>% gather(Method, Value, -n, -dist, -rep, -seed) %>%
  spread(dist, Value)
kable(Res) %>%
  kable_styling(bootstrap_options = "striped", full_width = F)
```

